python/triton/compiler/code_generator.py

```diff
@@ -624,9 +624,46 @@ def visit_If(self, node):
     def visit_IfExp(self, node):
         cond = self.visit(node.test)
         if _is_triton_tensor(cond):
-            raise UnsupportedLanguageConstruct(
-                None, node,
-                "Triton does not support `if` expressions (ternary operators) with dynamic conditions, use `if` statements instead")
+            cond = cond.to(language.int1, _builder=self.builder)
+            # TODO: Deal w/ more complicated return types (e.g tuple)
+            with enter_sub_region(self):
+                ip, last_loc = self._get_insertion_point_and_loc()
+
+                then_block = self.builder.create_block()
+                self.builder.set_insertion_point_to_start(then_block)
+                then_val = language.core._to_tensor(self.visit(node.body), self.builder)
+                then_block = self.builder.get_insertion_block()
+
+                else_block = self.builder.create_block()
+                self.builder.set_insertion_point_to_start(else_block)
+                # do not need to reset lscope since
+                # ternary expressions cannot define new variables
+                else_val = language.core._to_tensor(self.visit(node.orelse), self.builder)
```

Comment:
What if there's no return value?
```
a() if cond else b()
```

Reply:
By default, the compiler throws an assertion error -- stating that "ternary expression cannot return void", I don't think it would be difficult to change this, so instead, the ternary expression just doesn't yield anything. I just didn't do it, because it would involve more code.

Reply:
However, I can support no return values if you want.

Reply:
```
ternary expression cannot return void
```
Why would there be such a constraint?

Reply:
Is the constraint even there before this PR?

Reply:
> However, I can support no return values if you want. (The main question is, what should I do for the last line  return language.core.tensor(if_op.get_result(0), ret_type), if there is no return type).

Maybe you can try to just `return` without varaibles

Reply:
Before this PR, the ternary expression didn't support dynamic values at all -- so I suppose the constraint was there.

(All the code I've added is only in the dynamic section of the ternary expression, anything evaluating compile-time known conditions has been left untouched).

Reply:
Done.